src/tools/tools/process.fs

```diff
@@ -1171,10 +1171,130 @@ type internal ThreadPoolWait internal (id:int) as this =
+                                                                if bExist then 
+                                                                    tuple := ret
+                                                                    true
+                                                                else
+                                                                    failedReason := "Unable to find continuation Object"
+                                                                    false
+                                                            else
+                                                                failedReason := "Unable to find handle"
+                                                                false ),
+                                                    ThreadPoolWait.MaxSpinMs )
+                if bFind then 
+                    let rwh, infoFunc, workRef, refStartContinuation = !tuple
+                    refStartContinuation := DateTime.UtcNow.Ticks 
+                    let continuation, unblockHandle = !workRef
+                    // Long executing continuation may be flagged for performance issue. 
+                    continuation() 
+                    if Utils.IsNotNull unblockHandle  then 
+                        // If there is an unblock handle, set it. 
+                        unblockHandle.Set() |> ignore  
+                    (!handleCollectionRef).TryRemove( continuationObject ) |> ignore 
+                    if (!handleCollectionRef).IsEmpty then 
+                        ThreadPoolWait.WaitHandleCollection.TryRemove( handle) |> ignore
+                    rwh.Unregister(null) |> ignore 
+                else
+                    failwith (sprintf "ThreadPoolWait.Callback, can't find the entry of waithandle and continuation object after %d ms, reason: %s " ThreadPoolWait.MaxSpinMs !failedReason )
+            with 
+            | ex -> 
+                Logger.LogF( LogLevel.Error, fun _ -> sprintf "ThreadPoolWait.Callback, exception: %A" ex )
+                reraise()
+        else
+            failwith "ThreadPoolWait.Callback, timeout path reached (impossible execution route)"
+    static member TryTerminateAll() = 
+        for pair0 in ThreadPoolWait.WaitHandleCollection do
+            ThreadPoolWait.TryRemove( pair0.Key ) |> ignore 
+        ()
+    /// Mark all functions associated with handle as cancelled, the continuation function will not execute
+    static member TryRemove (handle:WaitHandle) = 
+        let bExist, handleCollection = ThreadPoolWait.WaitHandleCollection.TryGetValue( handle )
+        if bExist then 
+            for pair in handleCollection do 
+                let rwh, infoFunc, workRef, refStartContinuation = pair.Value
+                // If this get executed, the actual work will be cancelled. 
+                Volatile.Write( workRef,  (( fun _ -> () ), null) )  
```

Comment:
This code seems rather unusual way to achieve cancellation. In what case, we would want to remove a handle without getting cancellation request?  I think the "callback" should take a CancellationToken as a input parameter, and do cooperative cancellation inside it.  We also put the cancellation into the collection. And we should further assume TryRemove will be called only if the cancellation is requested.   Here, we can check whether the cancellation has been requested. If not,w e should regard it as an error and do not try to remove the handle.